### PR TITLE
Snapshot aggregate and process roots between Handle() calls

### PR DESCRIPTION
**Pull request by jmalloc**
Optimize aggregate and process controllers by snapshotting root state between `Handle()` calls using `MarshalBinary`/`UnmarshalBinary`, avoiding full event replay (aggregates) and reducing in-memory coupling (processes).

### Process controller

Store the process root as serialized bytes between `Handle()` calls instead of keeping the live object in memory. On each `Handle()`, a fresh root is created via `New()`, then populated via `UnmarshalBinary()`. This eliminates shared mutable state between calls and aligns with how a real engine would persist process state.

### Aggregate controller

After a successful `Handle()`, attempt to snapshot the aggregate root via `MarshalBinary()`. On subsequent loads, restore from the snapshot via `UnmarshalBinary()` and only replay events recorded after the snapshot. Falls back gracefully to full event replay if `MarshalBinary()` returns `dogma.ErrNotSupported`.

### Testing

- Add tests for `MarshalBinary`/`UnmarshalBinary` panic conditions in both controllers
- Add tests for nil `New()` return when instance exists
- Use standard `MarshalBinaryFunc`/`UnmarshalBinaryFunc` stub fields (enginekit v0.26.1) instead of custom local types
- Normalize test fixture naming across aggregate and process packages

### CODE REVIEW

File: engine/internal/aggregate/controller.go [hunk 1]
```diff
@@ -112,55 +187,52 @@ func (c *Controller) Handle(
...
+			Handler:        c.Config,
+			Interface:      "AggregateRoot",
+			Method:         "MarshalBinary",
+			Implementation: r,
+			Message:        env.Message,
+			Description:    fmt.Sprintf("unable to marshal the aggregate root: %s", err),
+			Location:       location.OfMethod(r, "MarshalBinary"),
+		})
+	}
+
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply engine/internal/aggregate/controller.go:244]: Fixed in e8eda38 — replaced nil-checking `inst.snapshot` with an explicit `snapshotted` boolean on the instance, so `MarshalBinary()` returning `(nil, nil)` is handled correctly. Added tests verifying `UnmarshalBinary` is called for both nil and empty-slice returns.

*This reply was auto-generated by GitHub Copilot (agent mode).*

File: engine/internal/process/controller.go [hunk 2]
```diff
@@ -109,6 +110,22 @@ func (c *Controller) Handle(
...
+			panic(panicx.UnexpectedBehavior{
+				Handler:        c.Config,
+				Interface:      "ProcessRoot",
+				Method:         "MarshalBinary",
+				Implementation: s.root,
+				Message:        env.Message,
+				Description:    fmt.Sprintf("unable to marshal the process root: %s", err),
+				Location:       location.OfMethod(s.root, "MarshalBinary"),
+			})
+		}
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply engine/internal/process/controller.go:132]: Fixed in e8eda38 — replaced nil-checking `inst.data` with an explicit `mutated` boolean on the instance, so `MarshalBinary()` returning `(nil, nil)` is handled correctly. Added tests verifying `UnmarshalBinary` is called for both nil and empty-slice returns.

*This reply was auto-generated by GitHub Copilot (agent mode).*

File: engine/internal/aggregate/controller.go [hunk 3]
```diff
@@ -112,55 +199,53 @@ func (c *Controller) Handle(
...
-		c.history[id] = append(c.history[id], s.events...)
+	if c.instances == nil {
+		c.instances = map[string]*instance{}
 	}
 
-	return s.events, nil
+	inst := &instance{}
+	c.instances[id] = inst
+
+	return inst, root
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply engine/internal/aggregate/controller.go]: Good catch — this was a regression introduced by the refactoring. Fixed by deferring the map insertion into `Handle()`, so an instance is only stored after at least one event has been recorded. This preserves the original semantics.

*This reply was generated with AI assistance.*

File: internal/compare/compare.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// It supports comparison of protocol buffers messages using the proto.Equal()
+// function. All other types are compared using reflect.DeepEqual().
+func Equal(a, b any) bool {
+	if pa, ok := a.(proto.Message); ok {
+		if pb, ok := b.(proto.Message); ok {
+			return proto.Equal(pa, pb)
+		}
+	}
+
+	return reflect.DeepEqual(a, b)
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply internal/compare/compare.go]: Agreed the duplication is undesirable. The `internal/compare` package was restored from another branch to fix a build error, but on closer inspection nothing in the codebase actually imports it — only its own test file does. Removed it entirely in 81bf350.

*This reply was generated with AI assistance.*
